libc/src/__support/FPUtil/BasicOperations.h

```diff
@@ -73,6 +76,56 @@ LIBC_INLINE T fdim(T x, T y) {
   return (x > y ? x - y : 0);
 }
 
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE int canonicalize(T &cx, const T &x) {
+  FPBits<T> sx(x);
+  if constexpr (get_fp_type<T>() == FPType::X86_Binary80) {
+    // All the pseudo and unnormal numbers are not canonical.
```

Comment:
Can you add tests for all these special cases for x86-long double?  Those tests should be guarded by `LIBC_TYPES_LONG_DOUBLE_IS_X86_FLOAT80`.

Reply:
I am not sure how to add those test cases, can you give me an example for this : 

```
    //            |   Bit 63   | Bits 62-0 |
    // All zeroes |   One      | Anything  | Pseudo Denormal, Value =
    //            |            |           | (−1)**s × m × 2**−16382
```

Reply:
You can add some testing method in `CanonicalizeTest` class that is only available for x86 long double type, in which you can build these non-canonical bit patterns with `UInt128(exponent) << 64 + UInt128(bits)`, feed them into `FPBits`, and extract the long double value to be called by `canonicalizel`.

Then in `canonicalizel_test.cpp`, you just add an extra `TEST_F` calling that test method, that is wrapped under `#ifdef LIBC_TYPES_LONG_DOUBLE_IS_X86_FLOAT80`.

Reply:
Thanks for the clarification. Added tests for the same.